From 6b5613e84e9a377819de522726087c15e85ea932 Mon Sep 17 00:00:00 2001
From: David Sniatkiwicz <davidsn@marvell.com>
Date: Thu, 5 Jul 2018 15:49:11 +0300
Subject: [PATCH 0159/1239] arm64: dts: a8k: add regulator current-limiter to
 USB3

- The USB device should linked to current-limiter through
"current-limiter" DTS property.
- This patch adds the "fixed-regulator" node for the USB current limiter.
- This patch adds handling for "current-limiter" property inside the USB3
node in A7K/A8K device trees for turning on/off the current limiter.

Change-Id: I6a7e3b8cb3d888c738e634db23949fdaca8e8c8b
Signed-off-by: David Sniatkiwicz <davidsn@marvell.com>
---
 arch/arm/dts/armada-7040-pcac.dts  | 24 ++++++++++++++++++++++++
 arch/arm/dts/armada-70x0-db.dtsi   | 18 ++++++++++++++++++
 arch/arm/dts/armada-8040-mcbin.dts |  1 +
 arch/arm/dts/armada-80x0-db.dtsi   | 26 ++++++++++++++++++++++++++
 4 files changed, 69 insertions(+)

diff --git a/arch/arm/dts/armada-7040-pcac.dts b/arch/arm/dts/armada-7040-pcac.dts
index 2beaa73211..cd10b9ca88 100644
--- a/arch/arm/dts/armada-7040-pcac.dts
+++ b/arch/arm/dts/armada-7040-pcac.dts
@@ -32,6 +32,23 @@
 		device_type = "memory";
 		reg = <0x0 0x0 0x0 0x80000000>;
 	};
+
+	simple-bus {
+		compatible = "simple-bus";
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		reg_usb3h1_current_limiter: usb3-current-limiter1 {
+			compatible = "regulator-fixed";
+			pinctrl-names = "default";
+			pinctrl-0 = <&cp0_xhci_vbus_pins>;
+			regulator-name = "reg-usb3h1-current-limiter";
+			regulator-min-microvolt = <900000>;
+			regulator-max-microvolt = <900000>;
+			regulator-force-boot-off;
+			gpio = <&cp0_gpio1 14 GPIO_ACTIVE_HIGH>; /* GPIO[46] */
+		};
+	};
 };
 
 &ap_pinctl {
@@ -78,6 +95,7 @@
 		 * XSMI  [35-36]
 		 * I2C	 [37-38]
 		 * UART1 [40-41]
+		 * USB CURRENT-LIMIT	[46]
 		 * UART2 [50-51]
 		 */
 		/*   0   1   2   3   4   5   6   7   8   9 */
@@ -88,6 +106,11 @@
 		     0   0   0   0   0   0   0   7   0   7
 		     6   6   0   0   0   0   0   0   0   0
 		     0   0   0>;
+
+	cp0_xhci_vbus_pins: cp0-xhci-vbus-pins {
+		marvell,pins = < 46 >;
+		marvell,function = <0>;
+	};
 };
 
 &cp0_spi1 {
@@ -133,6 +156,7 @@
 };
 
 &cp0_usb3_1 {
+	vbus-supply = <&reg_usb3h1_current_limiter>;
 	status = "okay";
 };
 
diff --git a/arch/arm/dts/armada-70x0-db.dtsi b/arch/arm/dts/armada-70x0-db.dtsi
index 6237660267..3b75aebed7 100644
--- a/arch/arm/dts/armada-70x0-db.dtsi
+++ b/arch/arm/dts/armada-70x0-db.dtsi
@@ -39,13 +39,31 @@
 				gpio = <&expander0 1 GPIO_ACTIVE_HIGH>;
 			};
 
+			cp0_reg_usb3_current_limiter0:cp0_usb3_current_limiter@0 {
+				compatible = "regulator-fixed";
+				regulator-min-microamp = <900000>;
+				regulator-max-microamp = <900000>;
+				regulator-force-boot-off;
+				gpio = <&expander0 4 GPIO_ACTIVE_HIGH>;
+			};
+
+			cp0_reg_usb3_current_limiter1: cp0_usb3_current_limiter@1 {
+				compatible = "regulator-fixed";
+				regulator-min-microamp = <900000>;
+				regulator-max-microamp = <900000>;
+				regulator-force-boot-off;
+				gpio = <&expander0 5 GPIO_ACTIVE_HIGH>;
+			};
+
 			cp0_usb3_0: usb3@500000 {
 				vbus-supply = <&cp0_reg_usb3_vbus0>;
+				current-limiter = <&cp0_reg_usb3_current_limiter0>;
 				vbus-disable-delay = <500>;
 			};
 
 			cp0_usb3_1: usb3@510000 {
 				vbus-supply = <&cp0_reg_usb3_vbus1>;
+				current-limiter = <&cp0_reg_usb3_current_limiter1>;
 				vbus-disable-delay = <500>;
 			};
 		};
diff --git a/arch/arm/dts/armada-8040-mcbin.dts b/arch/arm/dts/armada-8040-mcbin.dts
index 66355ab176..82295a54e5 100644
--- a/arch/arm/dts/armada-8040-mcbin.dts
+++ b/arch/arm/dts/armada-8040-mcbin.dts
@@ -246,6 +246,7 @@
 
 &cp1_usb3_0 {
 	vbus-supply = <&reg_usb3h0_vbus>;
+	/delete-property/ current-limiter;
 	status = "okay";
 };
 
diff --git a/arch/arm/dts/armada-80x0-db.dtsi b/arch/arm/dts/armada-80x0-db.dtsi
index 3544f8623a..2411bea8ab 100644
--- a/arch/arm/dts/armada-80x0-db.dtsi
+++ b/arch/arm/dts/armada-80x0-db.dtsi
@@ -45,6 +45,21 @@
 				gpio = <&expander0 1 GPIO_ACTIVE_HIGH>;
 			};
 
+			cp0_reg_usb3_current_limiter0:cp0_usb3_current_limiter@0 {
+				compatible = "regulator-fixed";
+				regulator-min-microamp = <900000>;
+				regulator-max-microamp = <900000>;
+				regulator-force-boot-off;
+				gpio = <&expander0 4 GPIO_ACTIVE_HIGH>;
+			};
+
+			cp0_reg_usb3_current_limiter1: cp0_usb3_current_limiter@1 {
+				compatible = "regulator-fixed";
+				regulator-min-microamp = <900000>;
+				regulator-max-microamp = <900000>;
+				regulator-force-boot-off;
+				gpio = <&expander0 5 GPIO_ACTIVE_HIGH>;
+			};
 			/*
 			 * Even though this node used for enable/disable
 			 * cp1 usb vbus-supply, we use "cp0" prefix since
@@ -59,12 +74,22 @@
 				gpio = <&expander1 0 GPIO_ACTIVE_HIGH>;
 			};
 
+			cp0_reg_usb3_current_limiter2: cp0_usb3_current_limiter@2 {
+				compatible = "regulator-fixed";
+				regulator-min-microamp = <900000>;
+				regulator-max-microamp = <900000>;
+				regulator-force-boot-off;
+				gpio = <&expander1 4 GPIO_ACTIVE_HIGH>;
+			};
+
 			cp0_usb3_0: usb3@500000 {
 				vbus-supply = <&cp0_reg_usb3_vbus0>;
+				current-limiter = <&cp0_reg_usb3_current_limiter0>;
 				vbus-disable-delay = <500>;
 			};
 			cp0_usb3_1: usb3@510000 {
 				vbus-supply = <&cp0_reg_usb3_vbus1>;
+				current-limiter = <&cp0_reg_usb3_current_limiter1>;
 				vbus-disable-delay = <500>;
 			};
 		};
@@ -75,6 +100,7 @@
 
 			cp1_usb3_0: usb3@500000 {
 				vbus-supply = <&cp0_reg_usb3_vbus2>;
+				current-limiter = <&cp0_reg_usb3_current_limiter2>;
 				vbus-disable-delay = <500>;
 			};
 		};
-- 
2.29.0

